gcc/d/d-decls.cc

```diff
@@ -101,7 +101,38 @@ VarDeclaration::toSymbol()
 	  return csym;
 	}
 
-      csym = new Symbol();
+      // Use only same Symbol for VarDeclarations with same mangle
+      if (!this->mangleOverride && isDataseg () && !(storage_class & STCextern))
+	{
+	  static hash_map<nofree_string_hash, Symbol *> duplicate_map(251);
+	  const char *vname = mangle(this);
+	  bool existed;
+
+	  Symbol **entry = &duplicate_map.get_or_insert(vname, &existed);
+	  if (!existed)
+	    {
+	      csym = new Symbol();
+	      *entry = csym;
+	    }
+	  else
+	    {
+	      csym = *entry;
+	      if (!csym->Sloc.equals(loc))
+		{
+		  error ("Found different variables with identical mangle: %s"
+		    " at %s and %s", vname, loc.toChars(), csym->Sloc.toChars());
```

Comment:
> Then maybe frontend should be made consistent?

Probably. The relevant code is `func.c bool FuncDeclaration::overloadInsert(Dsymbol *s)`. There's a check for this, but it's commented, so it's probably not that simple:
```c++
#if 0
    /* Disable this check because:
     *  const void foo();
     * semantic() isn't run yet on foo(), so the const hasn't been
     * applied yet.
     */
    if (type)
    {   printf("type = %s\n", type->toChars());
        printf("fd->type = %s\n", fd->type->toChars());
    }
    // fd->type can be NULL for overloaded constructors
    if (type && fd->type &&
        fd->type->covariant(type) &&
        fd->type->mod == type->mod &&
        !isFuncAliasDeclaration())
    {
        //printf("\tfalse: conflict %s\n", kind());
        return false;
    }
#endif
```

I'll have a look at your other comments tomorrow.